engine/language-server/src/main/scala/org/enso/languageserver/libraries/ComponentGroupsValidator.scala

```diff
@@ -22,92 +22,100 @@ final class ComponentGroupsValidator {
-            if (modulesMap.contains(moduleReference)) {
-              Left(
-                ValidationError
-                  .DuplicatedComponentGroup(libraryName, moduleReference)
-              )
-            } else {
-              modulesMap += moduleReference -> componentGroup
-              Right(config)
-            }
-          }
-          .find(_.isLeft)
+    groupsMap: mutable.Map[GroupReference, ComponentGroup]
+  ): Validator = { libraryName => componentGroups =>
+    componentGroups.newGroups
+      .map { componentGroup =>
+        val groupReference =
+          GroupReference(libraryName, componentGroup.group)
+        if (groupsMap.contains(groupReference)) {
+          Left(
+            ValidationError
+              .DuplicatedComponentGroup(libraryName, groupReference)
+          )
+        } else {
+          groupsMap += groupReference -> componentGroup
+          Right(componentGroups)
+        }
       }
-      .getOrElse(Right(config))
+      .find(_.isLeft)
+      .getOrElse(Right(componentGroups))
   }
 
   private def validateComponentGroupExtendsNothing(
-    modulesMap: mutable.Map[ModuleReference, ComponentGroup]
-  ): Config => Either[ValidationError, Config] = { config =>
-    val libraryName = LibraryName(config.namespace, config.name)
-    config.componentGroups.toOption
-      .flatMap { componentGroups =>
-        componentGroups.extendedGroups
-          .map { extendedComponentGroup =>
-            if (modulesMap.contains(extendedComponentGroup.module)) {
-              Right(config)
-            } else {
-              Left(
-                ValidationError.ComponentGroupExtendsNothing(
-                  libraryName,
-                  extendedComponentGroup.module
-                )
-              )
-            }
-          }
-          .find(_.isLeft)
+    groupsMap: mutable.Map[GroupReference, ComponentGroup]
+  ): Validator = { libraryName => componentGroups =>
+    componentGroups.extendedGroups
+      .map { extendedComponentGroup =>
+        if (groupsMap.contains(extendedComponentGroup.group)) {
+          Right(componentGroups)
+        } else {
+          Left(
+            ValidationError.ComponentGroupExtendsNothing(
+              libraryName,
+              extendedComponentGroup.group
+            )
+          )
+        }
       }
-      .getOrElse(Right(config))
+      .find(_.isLeft)
+      .getOrElse(Right(componentGroups))
   }
 
-  private def runValidation[A, E](xs: Iterable[Either[E, A]])(
-    fs: A => Either[E, A]*
-  ): Iterable[Either[E, A]] =
-    fs.foldLeft(xs) { (xs, f) =>
-      xs.map {
-        case Right(a) => f(a)
-        case Left(e)  => Left(e)
-      }
+  private def runValidation[K, V, E](init: Iterable[(K, Either[E, V])])(
+    validators: K => V => Either[E, V]*
+  ): Iterable[(K, Either[E, V])] =
+    validators.foldLeft(init) { (acc, validator) =>
+      acc.map { case (k, ev) => k -> ev.flatMap(validator(k)) }
```

Comment:
I think a comment explaining the intent behind this function would be useful. 
Currently it requires quite careful read to understand what it is meant to do. (I think especially the fact that we fold over validators whose type does not look sequential (it is, but it is very easy to miss the `*`) and not the init which is clearly `Iterable`.
Of course it makes sense after understanding it, but it took my a minute, so I think it would be worth to have some short description that would help in deciphering it.